### PR TITLE
Avoid using numeric_limits for sequence numbers and lock times

### PR DESCRIPTION
**Pull request by ryanofsky**
Switches to named constants, because numeric_limits calls can be harder to read and less portable.

Change was suggested by jamesob in https://github.com/bitcoin/bitcoin/pull/10973#discussion_r213473620

There are no changes in behavior except on some platforms we don't support (ILP64, IP16L32, I16LP32), where `SignalsOptInRBF` and `MutateTxAddInput` functions would now work correctly.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Tagged "refactoring" because the platforms that are broken are supposedly not supported.

**Comment by practicalswift** [comment]
utACK 8041cd39f0979a7db5d51bd1fc83d97d311836f7

File: src/policy/rbf.cpp [hunk 1]
```diff
@@ -7,7 +7,7 @@
 bool SignalsOptInRBF(const CTransaction &tx)
 {
     for (const CTxIn &txin : tx.vin) {
-        if (txin.nSequence < std::numeric_limits<unsigned int>::max()-1) {
+        if (txin.nSequence < CTxIn::SEQUENCE_FINAL - 1) {
```
Comment (hebasto) [inline src/policy/rbf.cpp]: could `if (txin.nSequence <= MAX_BIP125_RBF_SEQUENCE) {` ?

Comment (fanquake) [reply src/policy/rbf.cpp]: This was suggested by gmexwell [here](https://github.com/bitcoin/bitcoin/pull/14463#discussion_r224850088)

**Comment by fanquake** [comment]
@ryanofsky Can you combine the test change from #14463 into this? I'll close that PR, as this has more extensive changes.

**Comment by ryanofsky** [comment]
Updated 8041cd39f0979a7db5d51bd1fc83d97d311836f7 -> 90a9e728070c9dd6e236b036cc25ab9ff477f096 ([pr/climit.1](https://github.com/ryanofsky/bitcoin/commits/pr/climit.1) -> [pr/climit.2](https://github.com/ryanofsky/bitcoin/commits/pr/climit.2)) adding commits from #14463

File: src/script/script.h [hunk 2]
```diff
@@ -38,6 +38,12 @@ static const int MAX_STACK_SIZE = 1000;
 // otherwise as UNIX timestamp.
 static const unsigned int LOCKTIME_THRESHOLD = 500000000; // Tue Nov  5 00:53:20 1985 UTC
 
+// Maximum nLockTime. Since a lock time indicates the last invalid timestamp, a
+// transaction with this lock time will never be valid unless lock time
+// checking is disabled (by setting all input sequence numbers to
+// SEQUENCE_FINAL).
+static const uint32_t LOCKTIME_MAX = 0xffffffff;
```
Comment (gmaxwell) [inline src/script/script.h]: 0xFFFFFFFFU  ? lest we immediately get another hysterically titled "prevent integer overflow" PR?

Comment (practicalswift) [reply src/script/script.h]: @gmaxwell Why would an `U` suffix be needed for an integer literal with a hexadecimal base?

**Comment by ryanofsky** [comment]
Updated 90a9e728070c9dd6e236b036cc25ab9ff477f096 -> 535203075e50eedef8f00852328f81f440233278 ([pr/climit.2](https://github.com/ryanofsky/bitcoin/commits/pr/climit.2) -> [pr/climit.3](https://github.com/ryanofsky/bitcoin/commits/pr/climit.3)) adding U suffix (doesn't hurt to be explicit).

**Comment by practicalswift** [comment]
@ryanofsky I think the `U` suffix suggestion was just a misunderstanding based on an incorrect assumption on how C++ handles integer literals.

More specifically the incorrect assumption that the integer literal of `4294967295` can be assumed to have the same type as the integer literal `0xffffffff`.

That assumption is false :-)

Proof:

```
[cling]$ #include <typeindex>
[cling]$ #include <typeinfo>
[cling]$ std::type_index(typeid(4294967295)) == std::type_index(typeid(0xffffffff))
(bool) false
[cling]$ 4294967295
(long) 4294967295
[cling]$ 0xffffffff
(unsigned int) 4294967295
```

**Comment by ryanofsky** [comment]
The u suffix isn't needed, but it's nice because it makes it more obvious that compiler will choose an unsigned type. (According to the "type of the literal" table in https://en.cppreference.com/w/cpp/language/integer_literal, the compiler without the u would choose smallest signed or unsigned type that holds the value.)

**Comment by MarcoFalke** [comment]
utACK 535203075e50eedef8f00852328f81f440233278

* Checked that e4dc39b3bc4f199c863a71235f7269f7f9e336de happens to produce the same `bitcoind` binaries on my arch
* Don't care about bafb921507761217a2e9013a42e0aa619f831ccf
* Checked that 535203075e50eedef8f00852328f81f440233278 happens to produce the same `bitcoin-tx` and `bitcoind` binaries on my arch.

**Comment by ken2812221** [comment]
utACK 535203075e50eedef8f00852328f81f440233278

**Pull request merged**